Userland/Utilities/tsort.cpp

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2022, Eli Youngs <eli.m.youngs@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/CharacterTypes.h>
+#include <AK/HashMap.h>
+#include <LibCore/ArgsParser.h>
+#include <LibCore/Stream.h>
+#include <LibCore/System.h>
+#include <LibMain/Main.h>
+
+enum NodeStatus {
+    NotSeen,
+    Seen,
+    Prioritized,
+};
+
+struct Node {
+    StringView name;
+    HashTable<Node*> ancestors;
+    NodeStatus status;
+
+    unsigned hash() const { return name.hash(); }
+};
+
+using NodeMap = HashMap<StringView, Node*>;
+using NodeStack = Vector<Node*>;
+
+static Node* get_or_create_node(NodeMap& map, StringView const key)
+{
+    Optional<Node*> maybe_child_node = map.get(key);
```

Comment:
There is a `HashMap::ensure(key)` function which will create a node only if it doesn't exist.

Also please avoid manual memory management, you can use `OwnPtr<Node>` (or even `NonnullOwnPtr<Node>` )to store nodes in the hash map, which will take care of deleting it (Now you are not freeing nodes at all). 

You can do something like this:

```c++
map.ensure(key, [&]() { return adopt_own(*new Node { ... }); });
```